packages/wallets/ledgerconnect/src/adapter.ts

```diff
@@ -0,0 +1,236 @@
+
+interface LedgerConnectWalletEvents {
+    connect(...args: unknown[]): unknown;
+    disconnect(...args: unknown[]): unknown;
+}
+
+interface LedgerConnectWallet extends EventEmitter<LedgerConnectWalletEvents> {
+    isLedgerConnect?: boolean;
+    publicKey?: PublicKey;
+    isConnected: boolean;
+    signTransaction(transaction: Transaction): Promise<Transaction>;
+    signAndSendTransaction(
+        transaction: Transaction,
+        options?: SendOptions
+    ): Promise<{ signature: TransactionSignature }>;
+    signAllTransactions(transactions: Transaction[]): Promise<Transaction[]>;
+    connect(): Promise<void>;
+    disconnect(): Promise<void>;
+}
+
+interface LedgerConnectWindow extends Window {
+    ledgerConnectSolana?: LedgerConnectWallet;
+}
+
+declare const window: LedgerConnectWindow;
+
+export interface LedgerConnectWalletAdapterConfig {}
+
+export const LedgerConnectWalletName = 'Ledger Connect' as WalletName;
+
+export class LedgerConnectWalletAdapter extends BaseSignerWalletAdapter {
+    name = LedgerConnectWalletName;
+    url = 'https://www.ledger.com';
+    icon =
+        'data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTYwIiBoZWlnaHQ9IjE2MCIgdmlld0JveD0iMCAwIDE2MCAxNjAiIGZpbGw9Im5vbmUiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyI+CjxyZWN0IHdpZHRoPSIxNjAiIGhlaWdodD0iMTYwIiByeD0iMTYiIGZpbGw9IndoaXRlIi8+CjxwYXRoIGQ9Ik05My4xNDgyIDExOS4yMDdWMTI1SDEzNVY5OC44NzY5SDEyOC45MDJWMTE5LjIwN0g5My4xNDgyWk05My4xNDgyIDMzVjM4Ljc5MkgxMjguOTAyVjU5LjEyMzFIMTM1VjMzSDkzLjE0ODJaTTc0LjAxMDQgNTkuMTIzMUg2Ny45MTI1Vjk4Ljg3NjlIOTUuNDE1M1Y5My42NTM5SDc0LjAxMDRWNTkuMTIzMVpNMjYgOTguODc2OVYxMjVINjcuODUxOFYxMTkuMjA3SDMyLjA5NzlWOTguODc2OUgyNlpNMjYgMzNWNTkuMTIzMUgzMi4wOTc5VjM4Ljc5Mkg2Ny44NTE4VjMzSDI2WiIgZmlsbD0iIzAwMDAwRCIvPgo8L3N2Zz4K';
+
+    private _connecting: boolean;
+    private _wallet: LedgerConnectWallet | null;
+    private _publicKey: PublicKey | null;
+    private _readyState: WalletReadyState =
+        typeof window === 'undefined' || typeof document === 'undefined'
+            ? WalletReadyState.Unsupported
+            : WalletReadyState.NotDetected;
+
+    constructor(_config: LedgerConnectWalletAdapterConfig = {}) {
+        super();
+        this._connecting = false;
+        this._wallet = null;
+        this._publicKey = null;
+        if (this._readyState !== WalletReadyState.Unsupported) {
+            const handler = (event: MessageEvent<any>) => {
+                if (typeof event.data === 'object' && event.data.__ledgerConnect_loaded) {
+                    if (this._readyState !== WalletReadyState.Installed) {
+                        this._readyState = WalletReadyState.Installed;
+                        this.emit('readyStateChange', this._readyState);
+                    }
+                    window.removeEventListener('message', handler);
+                }
+            };
+
+            window.addEventListener('message', handler);
```

Comment:
Hmm, adding an event listener here will cause a memory leak whenever the adapter is instantiated, because it's only removed when the event is fired. This can happen very easily in the context of React applications.